### PR TITLE
FUNNEL_COUNT Aggregation Function

### PR DESCRIPTION
**Pull request by dario-liberman**
PR for #10866 

Solution includes the main roaring bitmaps implementation as well as an optimization for sorted columns.
Unit tests are included for both aggregation strategies.
This was also manually tested in a cluster.

### CODE REVIEW

**Comment by kishoreg** [comment]
This is amazing.Thanks for contributing.

what will the query look like. Does any other database support this functionality or they mostly rely on window function..

Can we update the docs as well to document this feature. Numbers look impressive. Will be great to blog this with actual usecase

**Comment by dario-liberman** [comment]
I am happy to document the function and create a blog post, but I would recommend we first merge it in silently (as experimental lets say), have it there for a bit for folks to try out and then go full out if it looks good, we can probably make a deal out of it similar to how was done with gap fill support.

In terms of other databases:
- [Doris sequence_match](https://doris.apache.org/docs/dev/sql-manual/sql-functions/aggregate-functions/sequence_match)
- [Clickhouse sequence_match](https://clickhouse.com/docs/en/sql-reference/aggregate-functions/parametric-functions#sequencematchpatterntimestamp-cond1-cond2-)

Neither of the above however do correlation, the user of these functions has to group by the correlation column and aggregate the results in an outer query in order to get similar results to what is proposed here.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/FunnelCountAggregationFunction.java [hunk 1]
```diff
@@ -0,0 +1,512 @@
...
+import org.apache.pinot.segment.spi.AggregationFunctionType;
+import org.apache.pinot.segment.spi.index.reader.Dictionary;
+import org.roaringbitmap.RoaringBitmap;
+
+
+/**
+ * The {@code FunnelCountAggregationFunction} calculates the number of step conversions for a given partition column and
+ * a list of boolean expressions.
+ * <p>IMPORTANT: This function relies on the partition column being partitioned for each segment, where there are no
+ * common values within different segments.
```
Comment (chenboat) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/FunnelCountAggregationFunction.java:45]: within --> across?

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/FunnelCountAggregationFunction.java [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+ * Example:
+ *   SELECT
+ *    dateTrunc('day', timestamp) AS ts,
+ *    FUNNEL_COUNT(
+ *      STEPS(url = '/addToCart', url = '/checkout', url = '/orderConfirmation'),
+ *      CORRELATED_BY(user)
+ *    ) as step_counts
+ *    FROM user_log
+ *    WHERE url in ('/addToCart', '/checkout', '/orderConfirmation')
+ *    GROUP BY 1
```
Comment (chenboat) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/FunnelCountAggregationFunction.java:58]: the example query looks good. Can you also add an example table with some rows of data and the corresponding query result?

**Comment by dario-liberman** [comment]
Please find proposed documentation here: https://github.com/pinot-contrib/pinot-docs/pull/187

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/FunnelCountAggregationFunction.java [hunk 3]
```diff
@@ -0,0 +1,512 @@
...
+    return isSorted(blockValSetMap) ? _sortedAggregationStrategy : _bitmapAggregationStrategy;
+  }
+
+  private SegmentAggregationStrategy<?, List<Long>> getAggregationStrategyByAggregationResult(Object aggResult) {
+    return aggResult instanceof SortedAggregationResult ? _sortedAggregationStrategy : _bitmapAggregationStrategy;
+  }
+
+  enum Option {
+    STEPS("steps"),
+    //SEQUENCE_BY("sequenceby"),
```
Comment (chenboat) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/FunnelCountAggregationFunction.java:221]: remove commented line.

**Comment by Jackie-Jiang** [comment]
In order for this aggregation to work, does it require all the data to be partitioned by segments (i.e. all users show up in the same segment, and no user can be shared across segments)? That is the pre-requisite for `SEGMENT_PARTITIONED_DISTINCT_COUNT`

**Comment by chenboat** [comment]
> In order for this aggregation to work, does it require all the data to be partitioned by segments (i.e. all users show up in the same segment, and no user can be shared across segments)? That is the pre-requisite for `SEGMENT_PARTITIONED_DISTINCT_COUNT`

Yes. That is the pre-requisite to use the aggregation function. For realtime table, it needs the Kafka topic to be partitioned (eg., by user ids).

**Comment by dario-liberman** [comment]
I am working on other aggregation strategies that do not require partitioning - https://github.com/apache/pinot/compare/master...dario-liberman:pinot:funnel-strategies

**Comment by kishoreg** [comment]
> > In order for this aggregation to work, does it require all the data to be partitioned by segments (i.e. all users show up in the same segment, and no user can be shared across segments)? That is the pre-requisite for `SEGMENT_PARTITIONED_DISTINCT_COUNT`
> 
> Yes. That is the pre-requisite to use the aggregation function. For realtime table, it needs the Kafka topic to be partitioned (eg., by user ids).

this is probably not practical and we should consider fixing this. Even if the kafka topic is partitioned by the same user_id, there is not guarantee that all users will be part of same segment.

**Comment by kishoreg** [comment]
I also noticed that this argument is taking steps and correlate column arguments as another function but we did not add any transform functions. This is violating the principle that arguments should be either literal, identifier or another function. 

syntactical steps is a function but we are not handling it that way.. this probably works today but it will probably fail when we move to V2 and enforce stricter constraints. Can we please fix this soon cc @dario-liberman @chenboat

**Comment by dario-liberman** [comment]
> > > In order for this aggregation to work, does it require all the data to be partitioned by segments (i.e. all users show up in the same segment, and no user can be shared across segments)? That is the pre-requisite for `SEGMENT_PARTITIONED_DISTINCT_COUNT`
> > 
> > Yes. That is the pre-requisite to use the aggregation function. For realtime table, it needs the Kafka topic to be partitioned (eg., by user ids).
> 
> this is probably not practical and we should consider fixing this. Even if the kafka topic is partitioned by the same user_id, there is not guarantee that all users will be part of same segment.

I shared above a work in progress branch with more funnel count aggregation strategies, effectively equivalents to DISTINCTCOUNT, DISTINCTCOUNTBITMAP and DISTINCTCOUNTTHETASKETCH. These do not depend on partitioning.

The strategy equivalent to SEGMENTPARTITIONEDDISTINCTCOUNT we have here is just a first version. When the column is configured as partition column we only have the same users across time boundaries between segments, which when grouping over time (eg per hour) to see funnel trends, gives good enough approximations. In the future it should be possible to incorporate a partition level (or server level?) phase so that we aggregate differently between segments within the same partition and segments across partitions. I will need more time for that though, for now I am adding different strategies so we can use the right one for each use case, as it will also depend on the sessionization window desired.

**Comment by dario-liberman** [comment]
> I also noticed that this argument is taking steps and correlate column arguments as another function but we did not add any transform functions. This is violating the principle that arguments should be either literal, identifier or another function. 
> 
> syntactical steps is a function but we are not handling it that way.. this probably works today but it will probably fail when we move to V2 and enforce stricter constraints. Can we please fix this soon cc @dario-liberman @chenboat 

I can work on this, the two solutions I can think of is to either declare the STEPS and other sub-functions as simply retuning their parameters (eg. an array), or do without them altogether. I liked the syntax of using sub-functions as a means to have named parameters in gapfill, so I replicated it here.

**Comment by dario-liberman** [comment]
> I am working on other aggregation strategies that do not require partitioning - [master...dario-liberman:pinot:funnel-strategies](https://github.com/apache/pinot/compare/master...dario-liberman:pinot:funnel-strategies)

@kishoreg @chenboat - Finally had a chance to add tests and complete the PR for the remaining aggregation strategies - https://github.com/apache/pinot/pull/11092

> > > > In order for this aggregation to work, does it require all the data to be partitioned by segments (i.e. all users show up in the same segment, and no user can be shared across segments)? That is the pre-requisite for `SEGMENT_PARTITIONED_DISTINCT_COUNT`
> > > 
> > > 
> > > Yes. That is the pre-requisite to use the aggregation function. For realtime table, it needs the Kafka topic to be partitioned (eg., by user ids).
> > 
> > 
> > this is probably not practical and we should consider fixing this. Even if the kafka topic is partitioned by the same user_id, there is not guarantee that all users will be part of same segment.
> 
> I shared above a work in progress branch with more funnel count aggregation strategies, effectively equivalents to DISTINCTCOUNT, DISTINCTCOUNTBITMAP and DISTINCTCOUNTTHETASKETCH. These do not depend on partitioning.
> 
> The strategy equivalent to SEGMENTPARTITIONEDDISTINCTCOUNT we have here is just a first version. When the column is configured as partition column we only have the same users across time boundaries between segments, which when grouping over time (eg per hour) to see funnel trends, gives good enough approximations. In the future it should be possible to incorporate a partition level (or server level?) phase so that we aggregate differently between segments within the same partition and segments across partitions. I will need more time for that though, for now I am adding different strategies so we can use the right one for each use case, as it will also depend on the sessionization window desired.

Hopefully these aggregation strategies address the concerns here?

Regarding the other concern about sub-function arguments not being real transform functions, that will be addressed in a separate PR.

**Pull request merged**